src/engraving/rendering/score/systemlayout.cpp

```diff
@@ -353,19 +355,27 @@ System* SystemLayout::collectSystem(LayoutContext& ctx)
      * Now perform all operation to finalize system.
      * **********************************************************/
 
-    // Brake cross-measure beams
+    // Break cross-measure beams
     if (ctx.state().prevMeasure() && ctx.state().prevMeasure()->isMeasure()) {
         Measure* pm = toMeasure(ctx.mutState().prevMeasure());
         BeamLayout::breakCrossMeasureBeams(pm, ctx);
     }
 
-    // hide empty staves
+    // Hide empty staves
     hideEmptyStaves(system, ctx, ctx.state().firstSystem());
 
+    // Re-create shapes to account for newly hidden/unhidden staves
+    // (and for potential forgotten shape updates, for example in MeasureLayout::setRepeatCourtesiesAndParens)
```

Comment:
A better fix for this is probably desirable, but outside the scope of this PR. The main reason for moving these lines out of `hideEmptyStaves` is to be able to do early returns in that method, without resorting to ugly DEFERs.

Reply:
Another good thing to do would be to only recreate the shapes when necessary. For this, hideEmptyStaves could return a bool indicating if it did indeed hide any stave.